### PR TITLE
Convert some lists in admin_client_mock to sets.  

### PR DESCRIPTION
**Pull request by sears**
This is testing the waters for doing this in a broader way throughout the codebase

# Description

No (intended) functional change.

## Type of change

*Please select one of the options below.*

- Other

## Discussion

*Are there any design details that you would like to discuss further?*

## Testing

The unit tests still pass.  No further testing should be needed.

## Documentation

N/A

## Follow-up

The sets-stored-as-arrays pattern is fairly common in this code base, and makes it difficult to read (and verbose).  We should consider modifying more logic in the same way.  (The runtime performance impact should be negligible).

### CODE REVIEW

**Comment by sears** [comment]
Log entry from failed test.  This seems unrelated, since the PR only touches mocks + unit tests.
```
------------------------------
• [FAILED] [853.417 seconds]
Test Operator Velocity
/codebuild/output/src956803802/src/github.com/FoundationDB/fdb-kubernetes-operator/fdb-kubernetes-tests/tests/test_operator_velocity/operator_velocity_test.go:150
  a knob is changed and a single process in the primary region is partitioned
  /codebuild/output/src956803802/src/github.com/FoundationDB/fdb-kubernetes-operator/fdb-kubernetes-tests/tests/test_operator_velocity/operator_velocity_test.go:231
    [It] should roll out knob changes within expected time
    /codebuild/output/src956803802/src/github.com/FoundationDB/fdb-kubernetes-operator/fdb-kubernetes-tests/tests/test_operator_velocity/operator_velocity_test.go:269

  Timed out after 420.001s.
  Expected
      <bool>: false
  to be true
  In [It] at: /codebuild/output/src956803802/src/github.com/FoundationDB/fdb-kubernetes-operator/fdb-kubernetes-tests/tests/test_operator_velocity/operator_velocity_test.go:86
------------------------------
```

**Comment by johscheuer** [comment]
> Log entry from failed test. This seems unrelated, since the PR only touches mocks + unit tests.
> 
> ```
> ------------------------------
> • [FAILED] [853.417 seconds]
> Test Operator Velocity
> /codebuild/output/src956803802/src/github.com/FoundationDB/fdb-kubernetes-operator/fdb-kubernetes-tests/tests/test_operator_velocity/operator_velocity_test.go:150
>   a knob is changed and a single process in the primary region is partitioned
>   /codebuild/output/src956803802/src/github.com/FoundationDB/fdb-kubernetes-operator/fdb-kubernetes-tests/tests/test_operator_velocity/operator_velocity_test.go:231
>     [It] should roll out knob changes within expected time
>     /codebuild/output/src956803802/src/github.com/FoundationDB/fdb-kubernetes-operator/fdb-kubernetes-tests/tests/test_operator_velocity/operator_velocity_test.go:269
> 
>   Timed out after 420.001s.
>   Expected
>       <bool>: false
>   to be true
>   In [It] at: /codebuild/output/src956803802/src/github.com/FoundationDB/fdb-kubernetes-operator/fdb-kubernetes-tests/tests/test_operator_velocity/operator_velocity_test.go:86
> ------------------------------
> ```

The test failures happens because this commit is missing: https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1425 (a rebase would fix this).

File: controllers/admin_client_mock.go [hunk 1]
```diff
@@ -41,9 +41,9 @@ type mockAdminClient struct {
 	Cluster                                  *fdbv1beta2.FoundationDBCluster
 	KubeClient                               client.Client
 	DatabaseConfiguration                    *fdbv1beta2.DatabaseConfiguration
-	ExcludedAddresses                        []string
+	ExcludedAddresses                        map[string]struct{}
```
Comment (johscheuer) [inline controllers/admin_client_mock.go:44]:
```suggestion
	ExcludedAddresses                        map[string]fdbv1beta2.None
```

We have a wrapper around that empty struct https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/api/v1beta2/none.go#L23-L25 to represent that this means it's a "None". We should use this everywhere, there might be some places where we missed to refactor that.

File: controllers/admin_client_mock.go [hunk 2]
```diff
@@ -41,9 +41,9 @@ type mockAdminClient struct {
...
 	ReincludedAddresses                      map[string]bool
-	KilledAddresses                          []string
+	KilledAddresses                          map[string]struct{}
```
Comment (johscheuer) [inline controllers/admin_client_mock.go:46]:
```suggestion
	KilledAddresses                          map[string]fdbv1beta2.None
```

File: controllers/admin_client_mock.go [hunk 3]
```diff
@@ -54,7 +54,7 @@ type mockAdminClient struct {
 	incorrectCommandLines                    map[string]bool
 	maxZoneFailuresWithoutLosingData         *int
 	maxZoneFailuresWithoutLosingAvailability *int
-	knobs                                    []string
+	knobs                                    map[string]struct{}
```
Comment (johscheuer) [inline controllers/admin_client_mock.go:57]: I believe this one is actually a valid list, is there a reason to convert this into a map?

File: controllers/admin_client_mock.go [hunk 4]
```diff
@@ -397,24 +375,14 @@ func (client *mockAdminClient) IncludeProcesses(addresses []fdbv1beta2.ProcessAd
...
-		included := false
-		for _, address := range addresses {
-			if address.String() == excludedAddress {
-				included = true
-				client.ReincludedAddresses[address.String()] = true
-				break
-			}
+	for _, address := range addresses {
+		address := address.String()
+		_, found := client.ExcludedAddresses[address]
```
Comment (johscheuer) [inline controllers/admin_client_mock.go:380]: Most go samples use "ok" in this context but I'm fine with "found", that comment is only informational.

File: controllers/admin_client_mock.go [hunk 5]
```diff
@@ -733,7 +701,10 @@ func (client *mockAdminClient) GetCoordinatorSet() (map[string]struct{}, error)
 
 // SetKnobs sets the knobs that should be used for the commandline call.
 func (client *mockAdminClient) SetKnobs(knobs []string) {
-	client.knobs = knobs
+	client.knobs = make(map[string]struct{}, len(knobs))
```
Comment (johscheuer) [inline controllers/admin_client_mock.go:704]: The list approach seemed simpler to me than the map approach.

File: controllers/bounce_processes_test.go [hunk 6]
```diff
@@ -145,14 +141,18 @@ var _ = Describe("bounceProcesses", func() {
...
 				for _, address := range processGroupAddresses {
-					addresses = append(addresses, fmt.Sprintf("%s:4501", address))
+					addresses[fmt.Sprintf("%s:4501", address)] = struct{}{}
 				}
 			}
-			Expect(adminClient.KilledAddresses).To(ContainElements(addresses))
+			// NOTE: Previously, this didn't check for equality.
+			// It checked that KilledAddresses was a superset of addresses.
+			// This was changed in a refactoring, under the assumption the old
+			// behavior was a typo.
```
Comment (johscheuer) [inline controllers/bounce_processes_test.go:154]: For strings the `Equal` command will be fine in other places we used `HaveLen` and the `ContainElements` to check that all elements included. If I remember correctly we had some issued with the `Equal` statement and pointers.

**Review by johscheuer — CHANGES_REQUESTED** [review]
Thanks for the refactoring. I think most of this is great and makes the code better, just a question around the client.knobs and a pointer to `fdbv1beta2.None`.

File: controllers/admin_client_mock.go (hunk 3 above)
Comment (sears) [reply controllers/admin_client_mock.go:57]: FDB treats knobs as a set internally, right?  (I've never seen a knob that can be set multiple times, or examples of the knob parsing caring about the order of the knobs.)

File: controllers/admin_client_mock.go (hunk 1 above)
Comment (sears) [reply controllers/admin_client_mock.go]: fixed

**Comment by sears** [comment]
Personally, I'd rather have the tests be flaky than to miss bugs that involve iteration order.  If this change does cause the tests to be flaky, then it's a sign of a serious issue somewhere else in the code base.  Also, by moving more of the logic to use sets instead of lists, we should be reducing the ways in which the system might be dependent on iteration order.

**Review by johscheuer — APPROVED** [review]
> Personally, I'd rather have the tests be flaky than to miss bugs that involve iteration order. If this change does cause the tests to be flaky, then it's a sign of a serious issue somewhere else in the code base. Also, by moving more of the logic to use sets instead of lists, we should be reducing the ways in which the system might be dependent on iteration order.

I agree with that statement and we should make sure we are not testing some logic that requires a specific order of the mocked status.

Changes LGTM 👍 

Once you took a look at the failing e2e tests we can merge those changes. Since the PR is only touching mocked files I guess the failure is unrelated and is one of the failures we investigate.

**Comment by sears** [comment]
These e2e tests failed:
```
logs/test_operator.log
logs/test_operator_ha_upgrades.log
```
I see failures like this from test_operator:
```json
{
  "level": "error",
  "ts": 1674086166.2564495,
  "msg": "Reconciler error",
  "controller": "foundationdbcluster",
  "controllerGroup": "apps.foundationdb.org",
  "controllerKind": "FoundationDBCluster",
  "foundationDBCluster": {
    "name": "fdb-cluster-ajd29ej5",
    "namespace": "test-operator-497-y3c1k2ia"
  },
  "namespace": "test-operator-497-y3c1k2ia",
  "name": "fdb-cluster-ajd29ej5",
  "reconcileID": "720ac139-cc9d-46f3-9679-281fb4b08070",
  "error": "FoundationDB error code 1031 (Operation aborted because the transaction timed out)",
  "stacktrace": "sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"
}
```
and like this from test_opearator_ha_upgrades:
```json
{
  "level": "error",
  "ts": 1674085706.7520964,
  "logger": "fdbclient",
  "msg": "Error from FDB command",
  "namespace": "test-operator-ha-upgrades-497-gtbpxfwu",
  "cluster": "fdb-cluster-5rrijb66-primary",
  "code": 1,
  "stdout": "ERROR: Not enough processes exist to support the specified configuration\nType `configure FORCE <TOKEN...>' to configure without this check\n",
  "stderr": "",
  "error": "exit status 1",
  "stacktrace": "github.com/FoundationDB/fdb-kubernetes-operator/fdbclient.(*cliAdminClient).ConfigureDatabase\n\t/workspace/fdbclient/admin_client.go:359\ngithub.com/FoundationDB/fdb-kubernetes-operator/controllers.updateDatabaseConfiguration.reconcile\n\t/workspace/controllers/update_database_configuration.go:99\ngithub.com/FoundationDB/fdb-kubernetes-operator/controllers.(*FoundationDBClusterReconciler).Reconcile\n\t/workspace/controllers/cluster_controller.go:166\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"
}
{
  "level": "error",
  "ts": 1674085706.7522104,
  "logger": "controller",
  "msg": "Error in reconciliation",
  "namespace": "test-operator-ha-upgrades-497-gtbpxfwu",
  "cluster": "fdb-cluster-5rrijb66-primary",
  "subReconciler": "controllers.updateDatabaseConfiguration",
  "requeueAfter": 0,
  "error": "exit status 1",
  "stacktrace": "github.com/FoundationDB/fdb-kubernetes-operator/controllers.(*FoundationDBClusterReconciler).Reconcile\n\t/workspace/controllers/cluster_controller.go:180\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"
}
{
  "level": "error",
  "ts": 1674085706.7522876,
  "msg": "Reconciler error",
  "controller": "foundationdbcluster",
  "controllerGroup": "apps.foundationdb.org",
  "controllerKind": "FoundationDBCluster",
  "foundationDBCluster": {
    "name": "fdb-cluster-5rrijb66-primary",
    "namespace": "test-operator-ha-upgrades-497-gtbpxfwu"
  },
  "namespace": "test-operator-ha-upgrades-497-gtbpxfwu",
  "name": "fdb-cluster-5rrijb66-primary",
  "reconcileID": "866555f2-5729-43e3-b0c9-2b8dcda4b3f4",
  "error": "exit status 1",
  "stacktrace": "sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"
}
{
  "level": "debug",
  "ts": 1674085706.7523012,
  "logger": "events",
  "msg": "Normal",
  "object": {
    "kind": "FoundationDBCluster",
    "namespace": "test-operator-ha-upgrades-497-gtbpxfwu",
    "name": "fdb-cluster-5rrijb66-primary",
    "uid": "36acdb89-e7b1-4d62-84fb-d4a96a2f0fa0",
    "apiVersion": "apps.foundationdb.org/v1beta2",
    "resourceVersion": "534892191"
  },
  "reason": "ReconciliationTerminatedEarly",
  "message": "exit status 1"
}
```

**Comment by sears** [comment]
This unrelated PR hit the same errors in test_operator:  https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1474

I assume the HA failures are a resource issue.

**Pull request merged**